doc/design/tensor_design.md

```diff
@@ -0,0 +1,175 @@
+# Tensor Design
+
+
+## Allocation
+
+`Allocation` is a [RAII](http://en.cppreference.com/w/cpp/language/raii) class template, which is used to handle a piece of memory. 
+
+```cpp
+// Template parametr 'Device' can be 'CpuDevice' or 'GpuDevice'
+template <typename Device>
+class Allocation {
+public:
+    Allocation();
+    Allocation(size_t size, Device device);
+
+    // Creates a non-owned allocation
+    Allocation(void* ptr, size_t size, Device device);
+
+    ~Allocation();
+    //No copying!
```

Comment:
But macro cannot put in `namespace`, `DISABLE_COPY` will pollute the global namespace.